File: FluidSolver_Parallel/FluidSolver.cpp
#include "FluidSolver.h"
#include "precice/SolverInterface.hpp"
#include <iostream>
#include <mpi.h>

using namespace precice;
using namespace precice::constants;

int main(int argc, char** argv)
{

  std::cout << "Starting Fluid Solver..." << std::endl;
  if (argc != 5) {
    std::cout << std::endl;
    std::cout << "Fluid: Usage: mpiexec -np <#procs> " << argv[0] << " <configurationFileName> <N> <tau> <kappa>" << std::endl;
    std::cout << std::endl;
    std::cout << "N:     Number of mesh elements, needs to be equal for fluid and structure solver." << std::endl;
    std::cout << "tau:   Dimensionless time step size." << std::endl;
    std::cout << "kappa: Dimensionless structural stiffness." << std::endl;

    return -1;
  }

  MPI_Init(&argc, &argv);

  int domainSize, gridOffset, rankWorld, sizeWorld, chunkLength;
  double *grid, tau, kappa; // Declare dataset

  MPI_Comm myComm = nullptr;

  MPI_Comm_rank(MPI_COMM_WORLD, &rankWorld);
  MPI_Comm_size(MPI_COMM_WORLD, &sizeWorld);

  int color = rankWorld / 2;

  MPI_Comm_split(MPI_COMM_WORLD, color, rankWorld, &myComm);
  
  int rankGroup, sizeGroup;
  MPI_Comm_rank(myComm, &rankGroup);
  MPI_Comm_size(myComm, &sizeGroup);

  std::cout << "Hi from local rank " << rankGroup << " or global rank " << rankWorld << " of color " << color << std::endl;


if (color == 0)
{
    
  domainSize = atoi(argv[2]);
  tau = atof(argv[3]);
  kappa = atof(argv[4]);

  if ((domainSize + 1) % sizeGroup == 0) {
    chunkLength = (domainSize + 1) / sizeGroup;
    gridOffset = rankGroup * chunkLength;
  } else if (rankGroup < (domainSize + 1) % sizeGroup) {
    chunkLength = (domainSize + 1) / sizeGroup + 1;
    gridOffset = rankGroup * chunkLength;
  } else {
    chunkLength = (domainSize + 1) / sizeGroup;
    gridOffset = ((domainSize + 1) % sizeGroup) * ((domainSize + 1) / sizeGroup + 1) + (rankGroup - ((domainSize + 1) % sizeGroup)) * (domainSize + 1) / sizeGroup;
  }

  std::vector<double> pressure(chunkLength);
  std::vector<double> pressure_n(chunkLength);
  std::vector<double > crossSectionLength(chunkLength);
  std::vector<double > crossSectionLength_n(chunkLength);
  std::vector<double> velocity(chunkLength);
  std::vector<double> velocity_n(chunkLength);
  std::vector<int> vertexIDs(chunkLength);

  //fluidDataDisplay(pressure, chunkLength);

  std::string configFileName(argv[1]);
  std::string solverName = "FLUID";

  SolverInterface interface(solverName, rankGroup, sizeGroup, &myComm);
  interface.configure(configFileName);

  int meshID = interface.getMeshID("Fluid_Nodes");
  int pressureID = interface.getDataID("Pressure", meshID);
  int crossSectionLengthID = interface.getDataID("CrossSectionLength", meshID);

  int dimensions = interface.getDimensions();
  grid = new double[dimensions * chunkLength];
  
  for (int i = 0; i < chunkLength; i++) {
    pressure[i] = 0.0;
    pressure_n[i] = 0.0;
    crossSectionLength[i] = 1.0;
    crossSectionLength_n[i] = 1.0;
    velocity[i] = 1.0 / kappa;
    velocity_n[i] = 1.0 / kappa;
    for (int j = 0; j < dimensions; j++) {
      grid[i * dimensions + j] = j == 0 ? gridOffset + (double)i : 0.0;
    }
  }

  interface.setMeshVertices(meshID, chunkLength, grid, vertexIDs.data());

  interface.initialize();

  double t = 0.0;
  double dt = 0.01;

  if (interface.isActionRequired(actionWriteInitialData())) {
    interface.writeBlockScalarData(pressureID, chunkLength, vertexIDs.data(), pressure.data());
    interface.fulfilledAction(actionWriteInitialData());
  }

  interface.initializeData();

  if (interface.isReadDataAvailable()) {
    interface.readBlockScalarData(crossSectionLengthID, chunkLength, vertexIDs.data(), crossSectionLength.data());
  }

  while (interface.isCouplingOngoing()) {
    int convergenceCounter = 0;
    if (interface.isActionRequired(actionWriteIterationCheckpoint())) {
      interface.fulfilledAction(actionWriteIterationCheckpoint());
    }

     // Call "Solver"
    fluidComputeSolution(rankGroup, sizeGroup, domainSize, chunkLength, kappa, tau, 0.0, t+dt,
                         pressure.data(), pressure_n.data(), pressure.data(),
                         crossSectionLength.data(), crossSectionLength_n.data(),
                         velocity.data(), velocity_n.data());

    //fluidDataDisplay(pressure, chunkLength);
    //fluidDataDisplay(crossSectionLength, chunkLength);

    interface.writeBlockScalarData(pressureID, chunkLength, vertexIDs.data(), pressure.data());

    interface.advance(dt);

    interface.readBlockScalarData(crossSectionLengthID, chunkLength, vertexIDs.data(), crossSectionLength.data());

    if (interface.isActionRequired(actionReadIterationCheckpoint())) { // i.e. not yet converged
      interface.fulfilledAction(actionReadIterationCheckpoint());
      convergenceCounter++;
    } else {
      t += dt;
      for (int i = 0; i < chunkLength; i++) {
        pressure_n[i] = pressure[i];
        velocity_n[i] = velocity[i];
        crossSectionLength_n[i] = crossSectionLength[i];
      }
    }
  }

  delete (grid);
  interface.finalize();

} // if color
  MPI_Finalize();

  return 0;
}

File: PythonTube/FluidSolver.py
from __future__ import division, print_function
import os
import sys
import argparse
import configuration_file as config
from thetaScheme import perform_partitioned_implicit_trapezoidal_rule_step, perform_partitioned_implicit_euler_step
import numpy as np
import tubePlotting

import matplotlib
import matplotlib.pyplot as plt
import matplotlib.animation as manimation

from output import writeOutputToVTK

from mpi4py import MPI
import precice
from precice import *

parser = argparse.ArgumentParser()
parser.add_argument("configurationFileName", help="Name of the xml precice configuration file.", type=str)

try:
    args = parser.parse_args()
except SystemExit:
    print("")
    print("Did you forget adding the precice configuration file as an argument?")
    print("Try $python FluidSolver.py precice-config.xml")
    quit()

print("Starting Fluid Solver...")

configFileName = args.configurationFileName

N = config.n_elem
dx = config.L / N  # element length

print("N: " + str(N))

solverName = "FLUID"

print("Configure preCICE...")
interface = precice.Interface(solverName, 0, 1)
interface.configure(configFileName)
print("preCICE configured...")

dimensions = interface.get_dimensions()

velocity = config.velocity_in(0) * np.ones(N+1)
velocity_n = config.velocity_in(0) * np.ones(N+1)
pressure = config.p0 * np.ones(N+1)
pressure_n = config.p0 * np.ones(N+1)
crossSectionLength = config.a0 * np.ones(N+1)
crossSectionLength_n = config.a0 * np.ones(N+1)

plotting_mode = config.PlottingModes.VIDEO
output_mode = config.OutputModes.VTK
writeVideoToFile = False

if plotting_mode == config.PlottingModes.VIDEO:
    fig, ax = plt.subplots(1)
    if writeVideoToFile:
        FFMpegWriter = manimation.writers['imagemagick']
        metadata = dict(title='PuleTube')
        writer = FFMpegWriter(fps=15, metadata=metadata)
        writer.setup(fig, "writer_test.mp4", 100)

meshID = interface.get_mesh_id("Fluid_Nodes")
crossSectionLengthID = interface.get_data_id("CrossSectionLength", meshID)
pressureID = interface.get_data_id("Pressure", meshID)

vertexIDs = np.zeros(N+1)
grid = np.zeros([dimensions, N+1])

grid[0,:] = np.linspace(0, config.L, N+1)  # x component
grid[1,:] = 0  # y component, leave blank

interface.set_mesh_vertices(meshID, N+1, grid.flatten('F'), vertexIDs)

t = 0

print("Fluid: init precice...")
precice_tau = interface.initialize()

if interface.is_action_required(action_write_initial_data()):
    interface.write_block_scalar_data(pressureID, N+1, vertexIDs, pressure)
    interface.fulfilled_action(action_write_initial_data())

interface.initialize_data()

if interface.is_read_data_available():
    interface.read_block_scalar_data(crossSectionLengthID, N+1, vertexIDs, crossSectionLength)

crossSectionLength_n = np.copy(crossSectionLength)
velocity_n = config.velocity_in(0) * crossSectionLength_n[0] * np.ones(N+1) / crossSectionLength_n  # initialize such that mass conservation is fulfilled

print(crossSectionLength_n)

while interface.is_coupling_ongoing():
    # When an implicit coupling scheme is used, checkpointing is required
    if interface.is_action_required(action_write_iteration_checkpoint()):
        interface.fulfilled_action(action_write_iteration_checkpoint())

    velocity, pressure, success = perform_partitioned_implicit_euler_step(velocity_n, pressure_n, crossSectionLength_n, crossSectionLength, dx, precice_tau, config.velocity_in(t + precice_tau), custom_coupling=False)
    interface.write_block_scalar_data(pressureID, N+1, vertexIDs, pressure)
    interface.advance(precice_tau)
    interface.read_block_scalar_data(crossSectionLengthID, N+1, vertexIDs, crossSectionLength)

    if interface.is_action_required(action_read_iteration_checkpoint()): # i.e. not yet converged
        interface.fulfilled_action(action_read_iteration_checkpoint())
    else: # converged, timestep complete
        t += precice_tau
        if plotting_mode is config.PlottingModes.VIDEO:
            tubePlotting.doPlotting(ax, crossSectionLength_n, velocity_n, pressure_n, dx, t)
            if writeVideoToFile:            
                writer.grab_frame()
            ax.cla()
        velocity_n = np.copy(velocity)
        pressure_n = np.copy(pressure)
        crossSectionLength_n = np.copy(crossSectionLength)
        if output_mode is config.OutputModes.VTK:
            writeOutputToVTK(t, "fluid", dx, N+1, datanames=["velocity", "pressure", "crossSection"], data=[velocity_n, pressure_n, crossSectionLength_n])

print("Exiting FluidSolver")

if plotting_mode is config.PlottingModes.VIDEO and writeVideoToFile:
    writer.finish()

interface.finalize()
